### ISSUE TITLE
Set up QA test suite for stage & prod deployments

### ISSUE THREAD

**Issue by groovecoder** [groovecoder/push-dev-dashboard#200]
Maybe a single Selenium end-to-end test:
- Sign up/in with Firefox Account
- Add push application with fixture app & VAPID key
- Validate the VAPID key
- Send a push messages with the VAPID key
- Verify the push message shows up in the push app details page
### Push app "fixture"

In order to test the entire dashboard, the Selenium test will need access to a fixture push application. Options are:
- Use https://mozilla-services.github.io/WebPushDataTestPage/
  - Would need to control 2 tabs to preserve the key
- Use a servicerworke.rs VAPID recipe? https://github.com/mozilla/serviceworker-cookbook/issues/218
- Use a WordPress fixture/test site (http://ec2-54-152-75-103.compute-1.amazonaws.com/wordpress/)
- All of the above

**Comment by groovecoder** [comment]
@davehunt - what do you think about the above? Do we have other selenium tests that use or span multiple sites/pages?

**Comment by davehunt** [comment]
It's not common, and chaining multiple sites in an e2e test increases fragility as you'll be at the mercy of several services that you may not have control over. UI integration tests are already rather brittle due to latency and their dependency on the underlying DOM of the application under test.

I can't say I fully understand the context of this issue, but I would generally recommend only writing UI automation for sites you have full control over, and if you do have to interact with third party sites keep this to a minimum. Where possible use documented APIs instead.

For signing in via Firefox Accounts, we do actually maintain a [page object model](https://github.com/mozilla/fxapom) as this is used in a number of our automation suites.

**Comment by groovecoder** [comment]
@davehunt in that case, we may need to do a manual QA check on the stage & prod servers after each deploy. Is that alright with you - do we have someone in QA who can commit to that?

**Comment by groovecoder** [comment]
I'm going to go ahead and try to make a selenium test suite that uses https://mozilla-services.github.io/WebPushDataTestPage/ to end-to-end test the dashboard.

**Comment by groovecoder** [comment]
@davehunt, @pdehaan - I've got selenium tests started in the repo. How can I start making them run on our QA server? I assume we have a Jenkins server with Selenium that I can use?

**Comment by rpappalax** [comment]
@davehunt @groovecoder I agree with Dave that tests that are tied to a remote service can be brittle and aren't always ideal.  For verifying service deployments, however, the reason for being of many of our tests is to verify that the actual service is running properly (i.e. when getting staged).  If you know that your tests will be "flaky", then just be aware that you may have the occasional failure.  The rest would be up to you to know when a test is "too" flaky due to latency issues or tolerably flaky giving you what you want.  But getting back to Dave's point - if you don't need to tie your test to a remote service, that's always better.

**Comment by groovecoder** [comment]
Note to self: Start a `push-dev-dashboard/` directory in [services-test repo](https://github.com/mozilla-services/services-test) with `e2e-test/run` like https://github.com/mozilla-services/services-test/blob/dev/demo/e2e-test/run

**Comment by groovecoder** [comment]
Adding push-dashboard to services-test here: https://github.com/mozilla-services/services-test/pull/88

**Comment by groovecoder** [comment]
@rpappalax I just updated this branch with an expanded test that gets all the way thru:
1. Signing into the dashboard with an existing FxA
2. Creating a new push app with a VAPID key
3. Validating the VAPID key by signing a unique "aud" claim payload JWT

The last bit we could do would be to actually send a push message with the VAPID headers and then check that it shows up on the dashboard. But that tests way more than just the dashboard ... i.e., it tests:
- Dashboard
- autopush service
- Push Messages API

In addition, it means that the test will need to configure the Firefox browser used to use the dev/stage/prod autopush service.

Let's plan to get into that next week though ... there may be some parts of it we decide to skip or fake.

**Comment by groovecoder** [comment]
@rpappalax - the `test_e2e` module is now merged to master. I'm leaving this open because it needs https://bugzilla.mozilla.org/show_bug.cgi?id=1273260 to work automatically.
